test/client/sauce-labs-runner.js

```diff
@@ -0,0 +1,147 @@
+var Promise = require('promise'),
+    request = require('request');
+
+var CHECK_RESULTS_TIMEOUT = 1000 * 2;
+
+var QUnitTestRunner = module.exports = function (options) {
+    var browsers = options.browsers || [];
+
+    browsers = browsers.map(function (item) {
+        return [item.platform || '', item.browserName || '', item.version || ''];
+    });
+
+    this.options = {
+        username:         options.username || '',
+        key:              options.key || '',
+        build:            options.build || Date.now(),
+        browsers:         browsers,
+        testName:         options.testName || 'QUnit tests',
+        tags:             options.tags || ['master'],
+        urls:             options.urls || [],
+        tunnelIdentifier: options.tunnelIdentifier || Math.floor((new Date()).getTime() / 1000 - 1230768000).toString()
+    };
+};
+
+QUnitTestRunner.prototype._sendRequest = function (params) {
+    var sendRequest = Promise.denodeify(request);
+
+    return sendRequest(params)
+        .then(function (result) {
+            var statusCode = result.statusCode;
+            var body       = result.body;
+
+            if (statusCode !== 200) {
+                throw [
+                    'Unexpected response from the Sauce Labs API.',
+                    params.method + ' ' + params.url,
+                    'Response status: ' + statusCode,
+                    'Body: ' + JSON.stringify(body)
+                ].join('\n');
+            }
+
+            return body;
+        },
+        function (error) {
+            throw 'Could not connect to Sauce Labs API: ' + error.toString();
+        }
+    );
+};
+
+QUnitTestRunner.prototype.runTests = function () {
+    var runner = this;
+
+    var runTaskPromises = this.options.urls.map(function (url) {
+        return runner._startTask(runner.options.browsers, url)
+            .then(function (taskIds) {
+                var completeTaskPromises = taskIds.map(function (id) {
+                    return runner._completeTask(id);
+                });
+
+                return Promise.all(completeTaskPromises);
+            })
+            .then(function (result) {
+                return result;
+            });
+    });
+
+    return Promise.all(runTaskPromises)
+        .then(function (results) {
+            return results;
```

Comment:
You don't need `then` in such cases. I've seen such code in other places around. Please, fix it there.